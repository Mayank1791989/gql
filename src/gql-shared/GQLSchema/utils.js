/* @flow */
import {
  type TypeNode,
  type NamedTypeNode,
  type EnumValueDefinitionNode,
  type FieldDefinitionNode,
  getDirectiveValues,
  GraphQLDeprecatedDirective,
  Kind,
} from 'graphql';

export function getNamedTypeNode(typeNode: TypeNode): NamedTypeNode {
  let namedType = typeNode;
  while (
    namedType.kind === Kind.LIST_TYPE ||
    namedType.kind === Kind.NON_NULL_TYPE
  ) {
    namedType = namedType.type;
  }
  return namedType;
}

/**
 * Given a field or enum value node, returns the string value for the
 * deprecation reason.
 */
export function getDeprecationReason(
  node: EnumValueDefinitionNode | FieldDefinitionNode,
): ?string {
  const deprecated = getDirectiveValues(GraphQLDeprecatedDirective, node);
  return deprecated && (deprecated.reason: any);
}
