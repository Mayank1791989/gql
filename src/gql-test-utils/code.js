/* @flow */
import { dedent } from 'dentist';
import splitLines from 'gql-shared/splitLines';

export default function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
