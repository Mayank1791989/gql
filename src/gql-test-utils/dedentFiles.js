/* @flow */
import { dedent } from 'dentist';

type FilesMap = { [path: string]: string };

export function dedentFiles(files?: FilesMap): FilesMap {
  if (!files) {
    return {};
  }
  return Object.keys(files).reduce((acc, path) => {
    acc[path] = dedent(files[path]);
    return acc;
  }, {});
}
