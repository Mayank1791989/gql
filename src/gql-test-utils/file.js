/* @flow */
import os from 'os';
import fs from 'fs-extra';
import path from 'path';

export const TEST_TEMP_DIR = path.join(
  // NOTE: os.tmpdir doesnt return realpath
  // so causing issues on osx ci server
  fs.realpathSync(os.tmpdir()),
  'gql-test',
);

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, `test-${Date.now()}`);

  // @TOOD: should we emptyDirSync?
  // NOTE: ensure directory is empty
  // when tests are run in watch mode we want to make sure files not present from last run
  // fs.emptyDirSync(dir);

  Object.keys(files).forEach(file => {
    const content = files[file];
    fs.outputFileSync(path.join(dir, file), content);
  });
  return dir;
}

export function removeTestTempDir() {
  fs.removeSync(TEST_TEMP_DIR);
}

export function replaceTempDir(filepath: string, str: string): string {
  return path.relative(TEST_TEMP_DIR, filepath).replace(/test-\d+/u, str);
}
