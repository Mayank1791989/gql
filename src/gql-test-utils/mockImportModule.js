/* @flow */
export default function mockImportModule(
  moduleName: string,
  mockFactory: () => $FixMe,
) {
  jest.doMock('import-from', () => {
    return function importFrom(dir, moduleID) {
      if (moduleID === moduleName) {
        return mockFactory();
      }
      const importFromActual = jest.requireActual('import-from');
      return importFromActual(dir, moduleID);
    };
  });
}
