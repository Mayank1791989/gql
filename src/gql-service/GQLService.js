/* @flow */
import GQLSchemaService from 'gql-schema-service';
import GQLQueryService from 'gql-query-service';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';
import GQLSchema from 'gql-shared/GQLSchema';
import log, { type LogListener } from 'gql-shared/log';
import { Source } from 'graphql';

const logger = log.getLogger('gql');

import {
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
  type GQLPosition,
} from 'gql-shared/types';

import { type GQLError } from 'gql-shared/GQLError';

type Options = $ReadOnly<{|
  configDir?: string,

  // watch
  watchman?: boolean,
  watch?: boolean,
|}>;

type CommandParams = $ReadOnly<{|
  sourcePath: string,
  sourceText: string,
  position: GQLPosition,
|}>;

export type { Options as GQLServiceOptions };

export default class GQLService extends GQLBaseService {
  _config: GQLConfig;
  _watcher: GQLWatcher;

  _services: [GQLSchemaService, GQLQueryService] = [];

  constructor(_options: ?Options) {
    super();
    const options = _options || {};

    this._config = new GQLConfig({
      configDir: options.configDir || process.cwd(),
    });
    this._watcher = new GQLWatcher({
      watchman: options.watchman,
      watch: options.watch,
    });

    // setup schema service
    const schemaService = new GQLSchemaService({
      config: this._config,
      watcher: this._watcher,
    });
    schemaService.onChange(this._triggerChange);
    schemaService.onError(this._triggerError);
    this._services[0] = schemaService;

    // setup query service
    if (this._config.getQueryConfig()) {
      const queryService = new GQLQueryService({
        config: this._config,
        schemaService,
        watcher: this._watcher,
      });
      queryService.onChange(this._triggerChange);
      queryService.onError(this._triggerError);
      this._services[1] = queryService;
    }
  }

  onLog(listener: LogListener) {
    return log.onLog(listener);
  }

  async _handleStart() {
    // start schema service
    await Promise.all(this._services.map(service => service.start()));
  }

  async _handleStop() {
    await Promise.all(this._services.map(service => service.stop()));
  }

  getSchema(): GQLSchema {
    const [schemaService] = this._services;
    return schemaService.getSchema();
  }

  getConfig(): GQLConfig {
    return this._config;
  }

  status(): Array<GQLError> {
    if (!this._isRunning) {
      return [];
    }
    try {
      return this._services.reduce((acc, service) => {
        return acc.concat(service.getErrors());
      }, []);
    } catch (err) {
      this._triggerError(err);
      return [];
    }
  }

  autocomplete(params: CommandParams): $ReadOnlyArray<GQLHint> {
    return this._catchThrownErrors(() => {
      logger.debug('autocomplete request');
      if (!this._isRunning) {
        return [];
      }

      logger.time('autocomplete response');
      const position = {
        line: params.position.line,
        column: params.position.column - 1,
      };
      const results = this._services.reduce((acc, service) => {
        acc.push(
          ...service.getHintsAtPosition({
            source: new Source(params.sourceText, params.sourcePath),
            position,
          }),
        );
        return acc;
      }, []);
      logger.timeEnd('autocomplete response');

      return results;
    }, []);
  }

  getDef(params: CommandParams): $ReadOnlyArray<GQLLocation> {
    return this._catchThrownErrors(() => {
      logger.debug('getDef request');
      if (!this._isRunning) {
        return [];
      }

      logger.time('getDef response');
      const results = this._services.reduce((acc, service) => {
        acc.push(
          ...service.getDefinitionAtPosition({
            source: new Source(params.sourceText, params.sourcePath),
            position: params.position,
          }),
        );
        return acc;
      }, []);
      logger.timeEnd('getDef response');

      return results;
    }, []);
  }

  findRefs(params: CommandParams): $ReadOnlyArray<GQLLocation> {
    return this._catchThrownErrors(() => {
      logger.debug('findRefs request');
      if (!this._isRunning) {
        return [];
      }

      logger.time('findRefs response');
      const results = this._services.reduce((acc, service) => {
        acc.push(
          ...service.findRefsOfTokenAtPosition({
            source: new Source(params.sourceText, params.sourcePath),
            position: params.position,
          }),
        );
        return acc;
      }, []);
      logger.timeEnd('findRefs response');

      return results;
    }, []);
  }

  getInfo(params: CommandParams): $ReadOnlyArray<GQLInfo> {
    return this._catchThrownErrors(() => {
      logger.debug('getInfo request');
      if (!this._isRunning) {
        return [];
      }

      logger.time('getInfo response');
      const result = this._services.reduce((acc, service) => {
        acc.push(
          ...service.getInfoOfTokenAtPosition({
            source: new Source(params.sourceText, params.sourcePath),
            position: params.position,
          }),
        );
        return acc;
      }, []);
      logger.timeEnd('getInfo response');

      return result;
    }, []);
  }
}
