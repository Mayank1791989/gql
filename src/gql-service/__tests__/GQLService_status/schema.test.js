/* @flow */
/* eslint-disable max-len */
import GQLService from '../../GQLService';
import { createTempFiles } from 'gql-test-utils/file';
import { status } from './utils';

test('should report errors in schema', async () => {
  const gql = new GQLService({
    configDir: createTempFiles({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/*.gql',
          }
        }
      `,
      'schema/schema.gql': `
        type Query {
          viewer: xViewer # missing viewer
        }
      `,
    }),
    watch: false,
  });
  gql.onError(err => {
    throw err;
  });
  await gql.start();
  expect(gql.status()).toMatchSnapshot();
  await gql.stop();
});

test('can modify validation rule severity', async () => {
  const gql = new GQLService({
    configDir: createTempFiles({
      'schema/schema.gql': `
        type Query {
          name: String
        }
        type Hello {
          name: String
        }
      `,
      '.gqlconfig': `{
        schema: {
          files: 'schema/*.gql',
          validate: {
            config: {
              NoUnusedTypeDefinition: 'error',
            },
          }
        }
      }`,
    }),
    watch: false,
  });
  gql.onError(err => {
    throw err;
  });
  await gql.start();
  expect(gql.status()).toMatchSnapshot();
  await gql.stop();
});

test('can turn off validation rules', async () => {
  const gql = new GQLService({
    configDir: createTempFiles({
      'schema/schema.gql': `
          type Query {
            name: String
          }
          type Hello {
            name: String
          }
        `,
      '.gqlconfig': `{
          schema: {
            files: 'schema/*.gql',
            validate: {
              config: {
                NoUnusedTypeDefinition: 'off',
              },
            }
          }
        }`,
    }),
    watch: false,
  });
  gql.onError(err => {
    throw err;
  });
  await gql.start();
  expect(gql.status()).toMatchSnapshot();
  await gql.stop();
});

test('report if no schema file exist or empty.', async () => {
  const gql = new GQLService({
    configDir: createTempFiles({
      '.gqlconfig': `{
          schema: {
            files: 'schema/*.gql',
          }
        }`,
    }),
    watch: false,
  });
  gql.onError(err => {
    throw err;
  });
  await gql.start();
  expect(gql.status()).toMatchSnapshot();
  await gql.stop();
});

test('Report if missing Query Root Type', async () => {
  const errors = await status({
    'schema/main.gql': `
      type Test {
        name: String!
      }
    `,
  });
  expect(errors).toMatchSnapshot();
});

describe('Report if Query type not ObjectType', () => {
  it('by type name', async () => {
    const errors = await status({
      'schema/main.gql': `
        input Query {
          name: String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  it('using schema', async () => {
    const errors = await status({
      'schema/main.gql': `
        scalar CustomScalar

        schema {
          query: CustomScalar
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });
});

test('Should able to use custom Query type', async () => {
  const errors = await status({
    'schema/main.gql': `
      type Viewer {
        id: ID!
      }

      schema {
        query: Viewer
      }
    `,
  });
  expect(errors).toEqual([]);
});

test('Report duplicate operation types', async () => {
  const errors = await status({
    'schema/main.gql': `
      type Viewer {
        id: ID!
      }
      type Player {
        name: String!
      }

      schema {
        query: Viewer
        query: Player
      }
    `,
  });
  expect(errors).toMatchSnapshot();
});

test('Report unknown operation type', async () => {
  const errors = await status({
    'schema/main.gql': `
      schema {
        query: Viewer
      }
    `,
  });
  expect(errors).toMatchSnapshot();
});

test('For simple schema only Query Type is enough', async () => {
  const errors = await status({
    'schema/main.gql': `
      type Query {
        name: String!
      }
    `,
  });
  expect(errors).toEqual([]);
});

describe('Report unknown types', () => {
  test('output field type', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }
        type Viewer {
          name: xString!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('input field type', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }
        input Viewer {
          name: xString!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('interface type', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        interface Node {
          id: ID!
        }

        type Viewer implements Node & xNode {
          id: ID!
          name: String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('union type', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        type Player {
          id: ID!
        }

        union Viewer = Player | xPlayer
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('arg type', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        type Player {
          image(size: xInt): String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });
});

describe('Report invalid objectType', () => {
  test('duplicate fields', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        type Player {
          id: ID!
          name: String!
          id: String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('field type not output type', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        type Player {
          name: String!
          value: InputValue!
          #--------^
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  // Bug in graphql-js validate not causing this error
  // enable test when fixed in graphql-js
  // eslint-disable-next-line no-restricted-properties
  test.skip('duplicate field args', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        type Player {
          image(size: Int!, id: String!, size: String): String!
        }
      `,
    });
    expect(errors).not.toEqual([]);
  });

  test('field arg type not input type', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        type Viewer {
          name: String!
        }

        type Player {
          image(size: Viewer!): String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('implements same interface multiple times', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        interface Node {
          id: ID!
        }

        interface User {
          id: ID!
          name: String!
        }

        type Player implements Node & User & Node {
          id: ID!
          name: String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('implementing non interface type', async () => {
    const errors = await status({
      'schema/main.gql': `
        type Query {
          name: String!
        }

        interface Node {
          id: ID!
        }

        type Player implements Node & String {
          id: ID!
          name: String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('type not implementing all interface fields', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/interface.gql': `
        interface User {
          id: ID!
          name: String!
        }
      `,
      'schema/main.gql': `
        type Player implements User {
          id: ID!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('implemented interface fields type mismatch', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/interface.gql': `
        interface User {
          id: ID!
          name: String!
        }
      `,
      'schema/main.gql': `
        type Player implements User {
          id: ID!
          name: String
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('type doesnt implements all interface field args', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/interface.gql': `
        interface User {
          id: ID!
          image(size: Int!): String!
        }
      `,
      'schema/main.gql': `
        type Player implements User {
          id: ID!
          image: String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('implemented interface fields args type mismatch', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/interface.gql': `
        interface User {
          id: ID!
          image(size: Int!): String!
        }
      `,
      'schema/main.gql': `
        type Player implements User {
          id: ID!
          image(size: String!): String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('type provides additional args for interface fields', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/interface.gql': `
        interface User {
          id: ID!
          image(size: Int!): String!
        }
      `,
      'schema/main.gql': `
        type Player implements User {
          id: ID!
          image(size: Int!, id: Int!): String!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });
});

describe('Report invalid unionType', () => {
  test('duplicate union members', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/interface.gql': `
        type User {
          id: ID!
          image(size: Int!): String!
        }
        type Player {
          id: ID!
        }
      `,
      'schema/main.gql': `
        union Viewer = User | Player | User
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('union member must be object type', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/types.gql': `
        type User {
          id: ID!
          image(size: Int!): String!
        }
      `,
      'schema/main.gql': `
        union Viewer = User | String
      `,
    });
    expect(errors).toMatchSnapshot();
  });
});

describe('Report invalid enumType', () => {
  test('duplicate values', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/main.gql': `
        enum Test {
          a
          b
          c
          a
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('invalid values', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/main.gql': `
        enum Test {
          true
          false
          null
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });
});

describe('Report invalid inputObjectType', () => {
  // enable when implemented in graphql-js
  // eslint-disable-next-line no-restricted-properties
  test.skip('duplicate fields', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/main.gql': `
        input PlayerCreateParams {
          id: ID!
          name: String!
          id: String!
        }
      `,
    });
    expect(errors).not.toEqual([]);
  });

  test('field type not inputType', async () => {
    const errors = await status({
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/main.gql': `
        type Player {
          name: String!
        }

        input PlayerCreateParams {
          id: ID!
          name: String!
          friends: [Player!]
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });
});

describe('Report unused type declaration', () => {
  test('Report unused type definition', async () => {
    const errors = await status({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/**/*.gql',
            validate: {
              config: {
                NoUnusedTypeDefinition: 'error'
              },
            }
          },
        }
      `,
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/main.gql': `
        type A {
          name: String
        }

        type X {
          name: A
          value: Int!
        }
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('Dont report type unused if type implements interface', async () => {
    const errors = await status({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/**/*.gql',
          },
        }
      `,
      'schema/query.gql': `
        type Query {
          name: String!
        }
      `,
      'schema/main.gql': `
        interface Node {
          id: String!
        }

        type X implements Node {
          id: String!
          value: Int!
        }
      `,
    });
    expect(errors).toEqual([]);
  });
});
