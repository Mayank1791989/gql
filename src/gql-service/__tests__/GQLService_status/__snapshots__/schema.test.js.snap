// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report duplicate operation types 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 10,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Must provide only one query type.",
    "severity": "error",
  },
]
`;

exports[`Report if Query type not ObjectType by type name 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 1,
        "line": 1,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Query root type must be Object type, it cannot be Query.",
    "severity": "error",
  },
]
`;

exports[`Report if Query type not ObjectType using schema 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 10,
        "line": 4,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Query root type must be Object type, it cannot be CustomScalar.",
    "severity": "error",
  },
]
`;

exports[`Report if missing Query Root Type 1`] = `
Array [
  Object {
    "locations": null,
    "message": "Query root type must be provided.",
    "severity": "error",
  },
]
`;

exports[`Report invalid enumType duplicate values 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
        "path": "$ROOT_DIR/schema/main.gql",
      },
      Object {
        "column": 3,
        "line": 5,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Enum type Test can include value a only once.",
    "severity": "error",
  },
]
`;

exports[`Report invalid enumType invalid values 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Enum type Test cannot include value: true.",
    "severity": "error",
  },
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 3,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Enum type Test cannot include value: false.",
    "severity": "error",
  },
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 4,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Enum type Test cannot include value: null.",
    "severity": "error",
  },
]
`;

exports[`Report invalid inputObjectType field type not inputType 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 12,
        "line": 8,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "The type of PlayerCreateParams.friends must be Input Type but got: [Player!].",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType duplicate fields 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 6,
        "path": "$ROOT_DIR/schema/main.gql",
      },
      Object {
        "column": 3,
        "line": 8,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Field Player.id can only be defined once.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType field arg type not input type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 15,
        "line": 10,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "The type of Player.image(size:) must be Input Type but got: Viewer!.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType field type not output type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 10,
        "line": 7,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Type \\"InputValue\\" not found.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType implemented interface fields args type mismatch 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 15,
        "line": 3,
        "path": "$ROOT_DIR/schema/interface.gql",
      },
      Object {
        "column": 15,
        "line": 3,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Interface field argument User.image(size:) expects type Int! but Player.image(size:) is type String!.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType implemented interface fields type mismatch 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 3,
        "path": "$ROOT_DIR/schema/interface.gql",
      },
      Object {
        "column": 9,
        "line": 3,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Interface field User.name expects type String! but Player.name is type String.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType implementing non interface type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 31,
        "line": 9,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Type Player must only implement Interface types, it cannot implement String.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType implements same interface multiple times 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 24,
        "line": 14,
        "path": "$ROOT_DIR/schema/main.gql",
      },
      Object {
        "column": 38,
        "line": 14,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Type Player can only implement Node once.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType type doesnt implements all interface field args 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 3,
        "path": "$ROOT_DIR/schema/interface.gql",
      },
      Object {
        "column": 3,
        "line": 3,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Interface field argument User.image(size:) expected but Player.image does not provide it.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType type not implementing all interface fields 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 3,
        "path": "$ROOT_DIR/schema/interface.gql",
      },
      Object {
        "column": 1,
        "line": 1,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Interface field User.name expected but Player does not provide it.",
    "severity": "error",
  },
]
`;

exports[`Report invalid objectType type provides additional args for interface fields 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 25,
        "line": 3,
        "path": "$ROOT_DIR/schema/main.gql",
      },
      Object {
        "column": 3,
        "line": 3,
        "path": "$ROOT_DIR/schema/interface.gql",
      },
    ],
    "message": "Object field argument Player.image(id:) is of required type Int! but is not also provided by the Interface field User.image.",
    "severity": "error",
  },
]
`;

exports[`Report invalid unionType duplicate union members 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 16,
        "line": 1,
        "path": "$ROOT_DIR/schema/main.gql",
      },
      Object {
        "column": 32,
        "line": 1,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Union type Viewer can only include type User once.",
    "severity": "error",
  },
]
`;

exports[`Report invalid unionType union member must be object type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 23,
        "line": 1,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Union type Viewer can only include Object types, it cannot include String.",
    "severity": "error",
  },
]
`;

exports[`Report unknown operation type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 10,
        "line": 2,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Specified query type \\"Viewer\\" not found.",
    "severity": "error",
  },
  Object {
    "locations": Array [
      Object {
        "column": 1,
        "line": 1,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Query root type must be provided.",
    "severity": "error",
  },
]
`;

exports[`Report unknown types arg type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 15,
        "line": 6,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Type \\"xInt\\" not found.",
    "severity": "error",
  },
]
`;

exports[`Report unknown types input field type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 5,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Type \\"xString\\" not found.",
    "severity": "error",
  },
]
`;

exports[`Report unknown types interface type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 31,
        "line": 9,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Type \\"xNode\\" not found.",
    "severity": "error",
  },
]
`;

exports[`Report unknown types output field type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 5,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Type \\"xString\\" not found.",
    "severity": "error",
  },
]
`;

exports[`Report unknown types union type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 25,
        "line": 9,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Type \\"xPlayer\\" not found.",
    "severity": "error",
  },
]
`;

exports[`Report unused type declaration Report unused type definition 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 1,
        "line": 5,
        "path": "$ROOT_DIR/schema/main.gql",
      },
    ],
    "message": "Unused type definition 'X' (NoUnusedTypeDefinition)",
    "severity": "error",
  },
]
`;

exports[`can modify validation rule severity 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 5,
        "path": "$ROOT_DIR/schema/schema.gql",
      },
    ],
    "message": "Unused type definition 'Hello' (NoUnusedTypeDefinition)",
    "severity": "error",
  },
]
`;

exports[`can turn off validation rules 1`] = `Array []`;

exports[`report if no schema file exist or empty. 1`] = `
Array [
  Object {
    "locations": null,
    "message": "No schema file found. Make sure schema file glob \\"schema/*.gql\\" matches atleast one file.",
    "severity": "error",
  },
  Object {
    "locations": null,
    "message": "Query root type must be provided.",
    "severity": "error",
  },
]
`;

exports[`should report errors in schema 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 19,
        "line": 3,
        "path": "$ROOT_DIR/schema/schema.gql",
      },
    ],
    "message": "Type \\"xViewer\\" not found.",
    "severity": "error",
  },
]
`;
